### ISSUE TITLE
ltsp.conf defaults

### ISSUE THREAD

**Issue by trickkiste** [ltsp/ltsp#56]
Currently default for variables settable in ltsp.conf are spread out throughout the whole code base. These should be consolidated into a single file for convenience.

One script checking if these vars are set in ltsp.conf and if not, it will assign the defaults.

I came to this conclusion talking with you about the BOOT_METHOD default for ltsp.conf. I would rather put the defaults directly into ltsp.conf, but you don't want to install an ltsp.conf in /etc, if the user does not want to change things.

Fair enough, but still these defaults should be set in an easy to find place and I would argue, the must not be mixed with code.

You could also read the default from /usr/share/ltsp/common/ltsp/ltsp.conf and either skip it completely if an /etc/ltsp/ltsp.conf is present, or overwrite the values from /usr/share (the default) with whatever is set in /etc.

That would be clean and easily readable.

And the defaults would always still be around if one has modified /etc/ltsp/ltsp.conf to a point that it has gotten totally non-functional. And they find everything in exactly the format, they need to work in, when they want to extend or modify the defaults.

**Comment by alkisg** [comment]
There is an infinite number of ltsp.conf parameters. For example, POST_INIT_RUN_THIS_COMMAND_PLEASE="...".
So it's not possible to initialize them all in a template ltsp.conf.

Another point is that there is a varying level of importance:

1. Some parameters are very significant should be there in a minimal template ltsp.conf (e.g. default setups)
2. Others of less significance should be documented in a more extensive ltsp.conf man page (e.g. advanced setups)
3. And a few are rare enough or not completely developed yet that should only be documented in the source code (to be used only after advice from the LTSP developers)

Additionally,  some are *autodetected* by default; for example, NAT is considered 1 if the server IP is 192.168.67.1; SSHFS is activated if /home/user doesn't exist at login time; etc. Some even depend on the value of other parameters for their default values. So it's not possible to initialize them with a bunch of var=value lines, as code is needed.

So thank you for the bug report but I'll close this with "won't fix"; if anyone does find ltsp.conf parameters that aren't appropriately documented wrt bullets 1-2-3 above, then that would be a bug.

**Comment by trickkiste** [comment]
As for the BOOT_METHOD parameter, having the defaults hidden somewhere in code, most certainly will lead to confusion. 

Maybe it would be a good idea to have a subset of these variables in such a central place - e.g. BOOT_METHOD.

**Comment by alkisg** [comment]
I'm not yet sure if BOOT_METHOD should be implemented or not; it depends on the answers to https://github.com/ltsp/ltsp/pull/51#issuecomment-546578492.

Even if it's implemented, I don't think that putting all initializations in the same place is a good idea. If you don't need BOOT_METHOD_AOE, why would you even show it to all LTSP users? If it's installed by an `apt install initramfs-aoe-method` package, why shouldn't the BOOT_METHOD be contained in that package?

Initially you said that you're in favor of feature separation and now you want all features centralized! OK enough chatting, let's get some work done, the Boot partition issue and a lot of others need coding time...

**Comment by trickkiste** [comment]
Hi Alkis,

no, I am not saying to include all possible scenarios in the default
variable initialization file.

What I am saying is, that you should provide BOOT_* methods in ltsp.conf
for the two boot methods you are currently using - convert them to use
BOOT_* methods.
And since you don't want to bless everybody with a default ltsp.conf file
in /etc/ltsp, the best course of action would be, to then source the
default ltsp.conf from /usr/share/ltsp to read the defaults from there (for
just the two active default methods).

This way, you don't have to install a default ltsp.conf but still have a
dedicated file to set ltsp.conf default values. Win-Win.

Hiding these defaults in code certainly makes life harder for every user.
You will end up with users asking, why they still have two BOOT methods
active, even though they only set one in their ltsp.conf ....
And the only way for them to find an answer would be, to dig through the
code.

I am actually trying to get work done as well.

I would consider the BOOT_* feature quite ready.
It is a minimal intervention, which opens up LTSP a great deal, so it can
support any given boot method.
It is not niche and the changes introduced don't change the outside
behaviour, as long as you keep the default variables.

Yes, maybe in spite of this change it would make sense to adapt `ltsp
info`, so it lists things by these BOOT_* methods now. I believe this
little change will not disrupt users all that much. If you believe people
rely on its output to stay exactly the same, I am sure there is a way to
allow for that also.

I would like to get this PR into upstream soon. So what do you want me to
change so you can accept it?

I am already testing the list you gave me. Don't want to answer them from
the top of my head, I am gathering prove.

The list of changes I will add shortly is as follows:
* BOOT_METHOD_* is too long, shorten to BOOT_*.
* Default BOOT_* should work with no ltsp.conf installed. Ergo, source
/usr/share/ltsp for defaults.
* Update & test `ltsp image`, `ltsp kernel`, `ltsp info` to work well with
BOOT_*

Anything else?
Feel free to name some things you want changed and I get on it.

Regards,
Markus

Am Mi., 30. Okt. 2019 um 11:45 Uhr schrieb Alkis Georgopoulos <
notifications@github.com>:

**Comment by trickkiste** [comment]
I will provide your answers.

And I am also very curious to understand, how these answers help you determine the worthiness of my PR?

But besides that, the BOOT_* method feature first and foremost provides a clean interface to extend LTSP to other boot methods, whatever they may be. 

Even the current methods used benefit from that, as things become easier to follow. And even if there is only 1% of users needing this, since there is no negative impact of this feature, doesn't it still make sense to provide it.

The RBD, iSCSI, AOE specific stuff, I agree, should then reside in their own respective initramfs packages - which would not hurt, if they were in LTSP PPA as well though.
